### ISSUE TITLE
SPI slave almost don't works while OTA update (IDFGH-3660)

### ISSUE THREAD

**Issue by saloid** [espressif/esp-idf#5595]
## Environment

- Development Kit:      none
- Module or chip used:  ESP32-WROVER-B
- IDF version (run ``git describe --tags`` to find it): v4.0.1
- Build System:         idf.py
- Operating System:     Windows
- (Windows only) environment type: Plain Command Prompt
- Using an IDE?: VSCode
- Power Supply:         external 3.3V

## Problem Description
SPI slave can't be used while OTA in progress. First five seconds (during flash erasing) interrupts almost completely disabled (purple line on screenshots - pin set inside interrupt handler). Later, while received firmware is writing to flash, interrupts are sometimes blocked too.

How can I increase SPI IRQ priority to prevent that? Or what can I do else?

### Steps to reproduce

1. Prepare spi slave configuration with regular spi transfers.
2. Run HTTPS OTA
3. You will miss many packets during OTA

## Screenshots
<details>
  <summary>Flash erasing 5 seconds block</summary>

![image](https://user-images.githubusercontent.com/22222390/87516708-7f302380-c686-11ea-943c-786f2dd9d304.png)
![image](https://user-images.githubusercontent.com/22222390/87516991-ee0d7c80-c686-11ea-9709-1614e7e5e8bf.png)

</details>
<details>
  <summary>IRQ blocked during further  flashing</summary>

![image](https://user-images.githubusercontent.com/22222390/87517467-8efc3780-c687-11ea-9769-ec1072dd692d.png)

</details>

**Comment by saloid** [comment]
Ok, looks like interrupts are really disabled during OTA update. But not all - interrupts from IRAM still enabled.
So I have marked my SPI hadler as IRAM interrupt: `esp_intr_set_in_iram(spihost[host]->intr, true);` (it already was in IRAM section), added `IRAM_ATTR` attribute to almost all functions, which are called from this IRQ.
And now I am getting `Guru Meditation Error: Core  0 panic'ed (Cache disabled but cached memory region accessed)` in first IRQ during OTA.

Backtrace shows that error was here: `0x4008b94e: get_rst_en_mask at .../components/driver/periph_ctrl.c:133`
<details>
<summary>Part of periph_ctrl.c</summary>

```c
static IRAM_ATTR uint32_t get_rst_en_mask(periph_module_t periph, bool enable)
{
    switch(periph) {    //! <-- line 133 is here
        case PERIPH_RMT_MODULE:
            return DPORT_RMT_RST;
        case PERIPH_LEDC_MODULE:
            return DPORT_LEDC_RST;
        case PERIPH_UART0_MODULE:
            return DPORT_UART_RST;
        case PERIPH_UART1_MODULE:
            return DPORT_UART1_RST;
        case PERIPH_UART2_MODULE:
            return DPORT_UART2_RST;
        case PERIPH_I2C0_MODULE:
            return DPORT_I2C_EXT0_RST;
        case PERIPH_I2C1_MODULE:
            return DPORT_I2C_EXT1_RST;
        case PERIPH_I2S0_MODULE:
            return DPORT_I2S0_RST;
        case PERIPH_I2S1_MODULE:
            return DPORT_I2S1_RST;
        case PERIPH_TIMG0_MODULE:
            return DPORT_TIMERGROUP_RST;
        case PERIPH_TIMG1_MODULE:
            return DPORT_TIMERGROUP1_RST;
        case PERIPH_PWM0_MODULE:
            return DPORT_PWM0_RST;
        case PERIPH_PWM1_MODULE:
            return DPORT_PWM1_RST;
        case PERIPH_PWM2_MODULE:
            return DPORT_PWM2_RST;
        case PERIPH_PWM3_MODULE:
            return DPORT_PWM3_RST;
        case PERIPH_UHCI0_MODULE:
            return DPORT_UHCI0_RST;
        case PERIPH_UHCI1_MODULE:
            return DPORT_UHCI1_RST;
        case PERIPH_PCNT_MODULE:
            return DPORT_PCNT_RST;
        case PERIPH_SPI_MODULE:
            return DPORT_SPI01_RST;
        case PERIPH_HSPI_MODULE:
            return DPORT_SPI2_RST;
        case PERIPH_VSPI_MODULE:
            return DPORT_SPI3_RST;
        case PERIPH_SPI_DMA_MODULE:
            return DPORT_SPI_DMA_RST;
        case PERIPH_SDMMC_MODULE:
            return DPORT_SDIO_HOST_RST;
        case PERIPH_SDIO_SLAVE_MODULE:
            return DPORT_SDIO_RST;
        case PERIPH_CAN_MODULE:
            return DPORT_CAN_RST;
        case PERIPH_EMAC_MODULE:
            return DPORT_EMAC_RST;
        case PERIPH_AES_MODULE:
            if (enable == true) {
                // Clear reset on digital signature & secure boot units, otherwise AES unit is held in reset also.
                return (DPORT_PERI_EN_AES | DPORT_PERI_EN_DIGITAL_SIGNATURE | DPORT_PERI_EN_SECUREBOOT);
            } else {
                //Don't return other units to reset, as this pulls reset on RSA & SHA units, respectively.
                return DPORT_PERI_EN_AES;
            }
        case PERIPH_SHA_MODULE:
            if (enable == true) {
                // Clear reset on secure boot, otherwise SHA is held in reset
                return (DPORT_PERI_EN_SHA | DPORT_PERI_EN_SECUREBOOT);
            } else {
                // Don't assert reset on secure boot, otherwise AES is held in reset
                return DPORT_PERI_EN_SHA;
            }
        case PERIPH_RSA_MODULE:
            if (enable == true) {
                // Also clear reset on digital signature, otherwise RSA is held in reset
                return (DPORT_PERI_EN_RSA | DPORT_PERI_EN_DIGITAL_SIGNATURE);
            } else {
                // Don't reset digital signature unit, as this resets AES also
                return DPORT_PERI_EN_RSA;
            }
        case PERIPH_WIFI_MODULE:
        case PERIPH_BT_MODULE:
        case PERIPH_WIFI_BT_COMMON_MODULE:
        case PERIPH_BT_BASEBAND_MODULE:
        case PERIPH_BT_LC_MODULE:
            return 0;
        default:
            return 0;
    }
}
```

</details>

As I can see from address `0x4008b94e`, this function successfully allocated in IRAM section (according to memory map).
<details>
<summary>Memory map</summary>

![image](https://user-images.githubusercontent.com/22222390/87551757-79a00100-c6b9-11ea-9a5e-de31b63715c9.png)

</details>

What else can I do?

**Comment by saloid** [comment]
Also I should add that I am using SPI slave in DMA mode and "DMA workaround" functions inside SPI IRQ were caused that crash.

**Comment by saloid** [comment]
Ok, finally I got it. So how to use spi slave during OTA update (I think, this manual will be applicable to any other hardware interrupts):
* Don't use DMA! Functions to work with DMA are located inside flash, which is inaccessible during OTA. So in this case you are limited to 32 byte data transfer in single SPI transaction.
* Move your SPI IRQ handler to core 1. To do it just launch `spi_slave_initialize()` and others from task, which is  [pinned to core 1](https://docs.espressif.com/projects/esp-idf/en/v4.0.1/api-reference/system/freertos.html#_CPPv423xTaskCreatePinnedToCore14TaskFunction_tPCKcK8uint32_tPCv11UBaseType_tPC12TaskHandle_tK10BaseType_t). This is needed to prevent interrupt blocking by WiFi and bluetooth (they all are on core 0)
* Allocate your SPI SLAVE functions in IRAM. In sdkconfig:
```
CONFIG_SPI_SLAVE_IN_IRAM=y
CONFIG_SPI_SLAVE_ISR_IN_IRAM=y
```
* Don't forget to add `IRAM_ATTR` attribute to your callback and to all functions, which are used inside.
* Add `.intr_flags=ESP_INTR_FLAG_IRAM` to `spi_bus_config` struct, which is passed as argument 2 in `spi_slave_initialize()`. This will acknowledge driver that your IRQ is located in IRAM and can be safely called while flash is disabled.

Also I edited IRQ handler in driver (removed RTOS queue usage because I don't need it in my case), but not sure that it matters.
